src/bpf/bpfctl.c

```diff
@@ -0,0 +1,220 @@
+                        return 0;
+
+                if (streq(json_variant_string(m), "map")) {
+                        static const JsonDispatch dispatch_table[] = {
+                                { "name",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfMapData, name),  JSON_MANDATORY },
+                                { "type", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfMapData, type), JSON_MANDATORY},
+                                { "memlock", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfMapData, memlock), JSON_MANDATORY},
+                                {}
+                        };
+                        _cleanup_(bpf_map_data_done) BpfMapData p = {};
+
+                        r = json_dispatch(v, dispatch_table, 0, &p);
+                        if (r < 0)
+                                return r;
+
+                        r = table_add_many(map_table,
+                                   TABLE_STRING, p.name, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.type, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_STRING, p.memlock, TABLE_SET_ALIGN_PERCENT, 100);
+                } else {
+                        static const JsonDispatch dispatch_table[] = {
+                                { "name",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfProgramData, name),  JSON_MANDATORY },
+                                { "type", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, type), JSON_MANDATORY},
+                                { "memlock", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfProgramData, memlock), JSON_MANDATORY},
+                                { "tag", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfProgramData, tag), JSON_MANDATORY},
+                                { "map_names", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfProgramData, map_names), 0},
+                                { "run_time_ns", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, run_time_ns), JSON_MANDATORY},
+                                { "run_count", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, run_cnt), JSON_MANDATORY},
+                                {}
+                        };
+                        _cleanup_(bpf_prog_data_done) BpfProgramData p = {};
+
+                        r = json_dispatch(v, dispatch_table, 0, &p);
+                        if (r < 0)
+                                return r;
+
+                        log_info(" prog_name: %s, prog_type: %lu, memlock: %s, prog_tag: %s, prog_map_names: %s, run_time_ns: %lu, run_cnt: %lu", p.name, p.type, p.memlock, p.tag, p.map_names, p.run_time_ns, p.run_cnt);
+                }
+                // table_print(prog_table, NULL);
+        }
+        table_print(map_table, NULL);
+
+        return r;
```

Comment:
please clean up error handling, i.e. we generally have the rule that functions should log about all their errors themselves, or that they should propagate them all (so that the caller can log). "main-function-like" functions should typically do the former, "library-like" the latter. The line between them is not always clear, but I'd say this is the former, i.e. it should log about all errors. (logging at debug level is always ok, and not considered logging in this case).

right now you handle one error with "log_oom()" (i.e. with proper logging), others you handle with debug logging, and even others you don't handle at all.

hence, please clean thus up, which probably means adding log_error_errno(…) or so everywhere